docs: document API layers (#1800)
opendp/opendp: docs/source/getting-started/quickstart.rst

            :start-after: demo
            :end-before: /demo

This demonstrates a number of low-level OpenDP patterns:

* First, define your "metric space": a data domain and a definition of distance.
* Then, chain operators together to construct a ``Measurement`` (aka mechanism).

> * Invoke that measurement on a value to get a DP release.

Reviewer: Not blocking, but for the future:
```suggestion
* Invoke that measurement on a dataset to get a DP release.
```

            :end-before: /demo

> This is obviously not the easiest way to add noise to a number,

Author: ```suggestion
```

> but it demonstrates a number of low-level OpenDP patterns:

Reviewer: I might drop "This is obviously not the easiest way to add noise to a number," unless you have an affinity to it.

Author: will do!

Author: Now:
> This demonstrates a number of low-level OpenDP patterns:

> * First, define your metric space.

Reviewer: Your prose is better, as it talks about the concepts and not the syntax.

Maybe another tweak though:
```suggestion
* First, define your metric space (your data domain and definition of neighboring)
```

Author: "definition of neighboring" is kind of clunk... Let me hold off on this and re-read the whole thing with updates.

Author: Ended up with 
> First, define your “metric space”: a data domain and a definition of distance

> * Then, chain operators together and end with a ``Measurement`` function.

Reviewer: ```suggestion
* Then, chain operators together to construct a ``Measurement`` (aka mechanism).
```

Author: Are we still planning to make a bulk change from "measurement" to "mechanism" at some point?

Reviewer: Yes, at some point.

* Invoke that measurement on a value to get a DP release.

OpenDP has two APIs and we'll demonstrate how to use both:

* The **Context API** is simpler and helps to enforce best practices. Currently available only for Python.

> * The **Framework API** is lower-level. Available for Python and R, it mirrors the underlying Rust framework.

Reviewer: ```suggestion
* The **Framework API** is lower-level. Available for Python, R and Rust, it directly implements the :ref:`OpenDP Programming Framework <programming-framework>`.
```

Because the Context API is a wrapper around the Framework API, it is easier to use but less flexible:
All calls ultimately pass through the Framework API.

> The next page will use the Framework and Context APIs to demonstrate the similarities between the Framework APIs in different languages.

Reviewer: ```suggestion
The next page will demonstrate usage of the Context API in Python, and Framework API in Python and R.
```

            :end-before: /demo

This is obviously not the easiest way to add noise to a number,
but it demonstrates a number of low-level OpenDP patterns:

* Defining your metric space with ``space_of`` in Python's Context API, or a (domain, distance) tuple in any language..
* Chaining operators together with ``>>`` in Python and Rust, or ``|>`` in R.
* Constructing a ``Measurement`` function on your metric space with ``then_laplace``.

> * Invoking that measurement on a value to get a DP release.

Author: Since these are a sequence of steps, it seems weird to move some but not all of this? I think I like it outside because the main point is to emphasize the similarities across languages.

OpenDP has layered APIs which provide increasing abstraction and usability:

* The **Framework API** is low-level. Available for Python and R, it mirrors the underlying Rust framework.
* The **Context API** introduces a ``Context`` class which ensures that queries do not exceed the privacy budget. Currently available only for Python.

> * The **Polars API** provides a DP extension to the `Polars <https://docs.pola.rs/>`_ dataframe library. Currently available only for Python.

Reviewer: I think it would be better to just state that we have two APIs: the Context API, and the Framework API. I'd list the Context API first. 

```
* The **Context API** is simpler to use and enforces best practices. The API is designed to ensure that mistakes that lead to under-estimated privacy guarantees cannot be made. Currently available only for Python.
* The **Framework API** is lower-level, and directly follows the OpenDP programming framework. Available for Python, R and Rust.
```

Polars and Sklearn cross-cut these: we aim to make either accessible from either, and we'll probably expand this set. I'd wait to introduce them until the tabular data and statistical modeling sections, respectively.

Author: So, essentially restore the original text here, but it still makes sense to move it up to quickstart -- I feel like Polars at least should be mentioned here as well, but maybe it doesn't need to be elevated to a bullet point?

Reviewer: That works for me.
